### ISSUE TITLE
Implement Tagging Feature

### ISSUE THREAD

**Issue by rmdstudio** [joomlatools/joomlatools-todo#7]

**Comment by rmdstudio** [comment]
Just a heads up that I am implementing tagging in com_todos first and then moving a lot of the tagging code to the com_tags afterwards. So don’t panic. I figured if I start from child component and then build up the parent, things will make sense much quicker.

**Comment by rmdstudio** [comment]
A lot of tags MVC has been removed from the com_todos and now it relies on com_tags. The administrator side is working, though I've been having difficulty getting site work. I am keep getting the error `Cannot load object from identifier: com://admin/todo.controller.behavior.taggable` I don't think it likes my bootstrapper. I'll find out what's wrong this week.

**Comment by rmdstudio** [comment]
I am trying to reuse com_tags layouts in com_todo without creating tags and tag views in com_todo. I assume that I need to create aliases in the bootstrapper.php but looks like the identifier that I create doesn’t pick up the layouts from com_tags. Does nooku fw even support that concept?

**Comment by johanjanssens** [comment]
@rmdstudio I will review your work tomorrow and help you get underway.

**Comment by rmdstudio** [comment]
@johanjanssens for now I am importing the com_tags layouts in com_todo layouts. It is less code, but it would have been nicer if we could have done without the tag(s) views in com_todo. Or if you are happy with the current approach then this issue is resolved.

**Comment by johanjanssens** [comment]
@rmdstudio The idea is that views live in the parent component and not in the re-usable component. The re-usable component could have views, but they should be extended/included in/from the components that composes it. 

This because we want template overrides to be specific to the component doing the composition, additionally we want tags to be application agnostic. It should work in Joomla, but also standalone or in Wordpress. The html it uses needs to be 100% generic.

**Comment by rmdstudio** [comment]
@johanjanssens @ercanozkaya this branch is ready to be reviewed. You need to install com_tags from this branch https://github.com/nooku/nooku-tags/tree/feature/1-fork before installing com_todo. I have also created sample data.

**Comment by ercanozkaya** [comment]
@rmdstudio  This part is looking good too. Do you have any work left on this or should we start merging?

**Comment by rmdstudio** [comment]
@ercanozkaya I need to use the tags in the front to filter tasks. That depends on the team consensus on what approach works best. For example:

1. adding tag filtering to existing menu item parameters and use them the same way categories are used to filter the results within the same itemid
2. ability to filter existing views by a specific tags. But what about tasks in different categories that use the same tag?
3. creating a specific view which shows a list of tags filtered by a tag
4. all of the above!

**Comment by rmdstudio** [comment]
I am adding @johanjanssens email reply here for my reference:

> In a first step I think we the following :

> 1. Users should be able to 'Filter' documents displayed in a menu item based on one or more tags that are available. This would result in a list of documents, displayed according to the rules of the menu item layout that adhere to those tags. 

> 2. It should be possible to filter the documents based on one or more tags in the administrator. For that we might need a filter to select the tags, can implement that in the sidebar. 

> Additionally, tags should be controllable by ACL, so that users can be prevents from adding additional tags, and instead just select them. Tags would then only be able to be added by people who have permissions to do so. This is important to prevent getting a huge set of tags which diffuse the use of them. 

> Next step : I would get started by adding a filter option to com_todo to see how filtering todo items based on tags would work. Then in a next step add the option to also filter the list of todo items by tag. 

cheers

**Comment by rmdstudio** [comment]
I am back to working on this issue

**Comment by rmdstudio** [comment]
We can now filter list of tasks using a list of tags in the menu item parameters. Next I will be developing filtering in the admin side.

**Comment by johanjanssens** [comment]
@rmdstudio  I have added some comments to the code, might have added comments on code you refactored later on, was a bit hard to see what code was added and later removed and what not.

**Comment by rmdstudio** [comment]
Building the sidebar tags filter in the admin side.

**Comment by rmdstudio** [comment]
@johanjanssens @ercanozkaya whether I use params or request in the xml file, I noticed that the tag values stay in the select list box. I think they are stored somewhere in the session. Before I dig deep inside the code, any tips on how to deal with that?

**Comment by rmdstudio** [comment]
I realized that I had a bad data. The link was saved as `index.php?option=com_todo&view=tasks&tag[0]=6` and I wasn't able to remove that tag. It is working now.

**Comment by rmdstudio** [comment]
@ercanozkaya @johanjanssens lots of code reading and making adjustments in the last 2 days.
 
1. Tags are now using slugs instead of ids
2. I am still using params for the menu items instead of request. Looks like request needs at least 1 value or am I wrong? One of my implementations had created a bad data and that wasted my time trying to debug it https://github.com/joomlatools/joomlatools-todo/issues/7#issuecomment-167736259
3. I am implementing the tags filter for the tasks list in the back-end. For some reason the tag query parameters stay in the session and accumulate. How would I prevent that? I couldn't find any code examples that helped grasp this. 

Thank you!

**Comment by rmdstudio** [comment]
<img width="963" alt="screen shot 2015-12-29 at 2 10 00 am" src="https://cloud.githubusercontent.com/assets/559971/12032821/52aa105a-add1-11e5-9d2c-e28ee138d46a.png">

Here is a screen capture. The filter happens, but when I remove the tags and resubmit the values are still in the session. What is the proper way to pass no tags so the tag state will unset? If you could just point me to a code example that will work. Thank you!

**Comment by ercanozkaya** [comment]
Hi @rmdstudio 

That is known to happen after emptying select2 boxes. This is what we use in LOGman to remedy the problem:
```
$('#activities-filter').on('submit', function(e) {
    if (!$('select[name="usergroup[]"]').select2("val").length) {
            $(this).append('<input type="hidden" name="usergroup" value="" />');
    }
});
```

About the menu parameter, no it doesn't require any values. Can you explain what happens when you move the parameter into request?

**Comment by rmdstudio** [comment]
@ercanozkaya thank you I'll try the js code. Also let me try the request menu parameters one more time now that I've got things working. I'll get back to you on this.

**Comment by rmdstudio** [comment]
@ercanozkaya so this actually worked:

```
<script>
kQuery(function($){

  $('#tasks-form').on('submit', function( event ) {
      if (!$('select[name="tag[]"]').select2("val").length) {
          $(this).find('input[name="tag[]"]').remove();
          $(this).append('<input type="hidden" name="tag" value="" />');
      }
  });

});
</script>
```

It is in the sidebar layout, but I think I better put it in a js file.

**Comment by rmdstudio** [comment]
<img width="1230" alt="screen shot 2015-12-29 at 10 27 38 pm" src="https://cloud.githubusercontent.com/assets/559971/12047155/de34c8ae-ae7b-11e5-8ff2-d942544bfdc4.png">

@ercanozkaya @johanjanssens 

The admin side tag filter is working. Do we want all the CSS work to show the filter in the sidebar or this will do as a proof of concept?

ps: going to test the params/request concept in the menu item now

**Comment by rmdstudio** [comment]
@ercanozkaya @johanjanssens  so this is what happens when I use _request_ instead of _params_: the tag[] array is added to the url as a query parameter which is expected also I can delete the last value in the tags select list. It wants to have at least one item in the list. Do you know how I can prevent that?

**Comment by rmdstudio** [comment]
@ercanozkaya @johanjanssens yes it was the same phenomenon. I fed it more javascript. Let me clean up the code and commit.

**Comment by rmdstudio** [comment]
@ercanozkaya @johanjanssens 

OK code is in. So what's left:

1. changing relations table name. I forgot what you requested and can't find the comment I'm afraid.
2. styling the tags filter in the admin side and putting it in the sidebar perhaps? That's just some CSS work? Or is it unnecessary?

**Comment by rmdstudio** [comment]
Also let me know when I should start with tags in DocMAN. I can get started on that right away.

**Comment by johanjanssens** [comment]
@rmdstudio About your questions :+1: 

1. I have added additional comments.

2. I would try to implement the tag filter in the sidebar. Check the LOGman Filter by Users filter, where you can select multiple groups for inspiration.

3. What do you mean with access control ?  We can discuss this on slack.

**Comment by rmdstudio** [comment]
Just committing my code. It's not done yet.

**Comment by rmdstudio** [comment]
Updated the code to work with the latest changes in the Nooku Tags

**Comment by rmdstudio** [comment]
Updated the install.sql queries and removed taggable behaviour from the task controllers.
